wp-includes/parser/class-wp-parser-node.php

```diff
@@ -102,83 +102,169 @@ public function merge_fragment( $node ) {
 		}
 		return null;
 	}
 
-	public function get_descendants( $rule_name ) {
-		$parse_trees     = array( $this );
+	public function get_children(): array {
+		return $this->children;
+	}
+
+	public function get_child_nodes( ?string $rule_name = null ): array {
+		$nodes = array();
+		foreach ( $this->children as $child ) {
+			if (
+				$child instanceof WP_Parser_Node
+				&& ( null === $rule_name || $child->rule_name === $rule_name )
+			) {
+				$nodes[] = $child;
+			}
+		}
+		return $nodes;
+	}
+
+	public function get_child_tokens( ?int $token_id = null ): array {
+		$tokens = array();
+		foreach ( $this->children as $child ) {
+			if (
+				$child instanceof WP_Parser_Token
+				&& ( null === $token_id || $child->id === $token_id )
+			) {
+				$tokens[] = $child;
+			}
+		}
+		return $tokens;
+	}
+
+	public function get_descendants(): array {
+		$nodes           = array( $this );
 		$all_descendants = array();
-		while ( count( $parse_trees ) ) {
-			$parse_tree      = array_pop( $parse_trees );
-			$all_descendants = array_merge( $all_descendants, $parse_tree->get_children( $rule_name ) );
-			array_push( $parse_trees, ...$parse_tree->get_children() );
+		while ( count( $nodes ) ) {
+			$node            = array_shift( $nodes );
+			$all_descendants = array_merge( $all_descendants, $node->get_children() );
+			$children        = $node->get_child_nodes();
+			if ( count( $children ) > 0 ) {
+				array_push( $nodes, ...$children );
+			}
 		}
 		return $all_descendants;
 	}
 
-	public function get_children( $rule_name = null ) {
-		$matches = array();
-		foreach ( $this->children as $child ) {
-			if ( $child instanceof WP_Parser_Node && (
-				null === $rule_name ||
-				$child->rule_name === $rule_name
-			) ) {
-				$matches[] = $child;
+	public function get_descendant_nodes( ?string $rule_name = null ): array {
+		$nodes           = array( $this );
+		$all_descendants = array();
+		while ( count( $nodes ) ) {
+			$node            = array_shift( $nodes );
+			$all_descendants = array_merge( $all_descendants, $node->get_child_nodes( $rule_name ) );
+			$children        = $node->get_child_nodes();
+			if ( count( $children ) > 0 ) {
+				array_push( $nodes, ...$children );
 			}
 		}
-		return $matches;
+		return $all_descendants;
 	}
+
+	public function get_descendant_tokens( ?int $token_id = null ): array {
+		$nodes           = array( $this );
+		$all_descendants = array();
+		while ( count( $nodes ) ) {
+			$node            = array_shift( $nodes );
+			$all_descendants = array_merge( $all_descendants, $node->get_child_tokens( $token_id ) );
+			$children        = $node->get_child_nodes();
+			if ( count( $children ) > 0 ) {
+				array_push( $nodes, ...$children );
+			}
+		}
+		return $all_descendants;
+	}
+
+	/*
+	 * @TODO: Let's implement a more powerful AST-querying API.
+	 *        See: https://github.com/WordPress/sqlite-database-integration/pull/164#discussion_r1855230501
```

Comment:
If we're refactoring the API, let's annotate each method we expect to change as `@private` or `@internal` and add a comment for the API consumers explaining the reasons not to use them yet. Otherwise, we'll introduce breaking changes without a prior notice. Maybe we could `@deprecate` them right away – that way the IDEs will provide visual hints.

Reply:
@adamziel Maybe, while we're behind the feature flag and have a bunch of `TODO`s, we could still consider all the new APIs to be experimental and subject to change? Given we're under the Automattic org repo, maybe that would be enough? Or do you think it's safer to annotate all public methods for now?

Reply:
It could be a separate ticket, we's just have to resolve it before handing the api over for consumption by another team